### ISSUE TITLE
[BUG] DMARC quarantine (add_header) v. greylist - no action taken

### ISSUE THREAD

**Issue by Player701** [rspamd/rspamd#4535]
### Prerequisites

* [X] Put an X between the brackets on this line if you have done all of the following:
    * Read about bug reporting in general: https://rspamd.com/doc/faq.html#how-to-report-bugs-found-in-rspamd
    * Enabled relevant debugging logs: https://rspamd.com/doc/faq.html#how-to-debug-some-module-in-rspamd 
    * Checked the FAQs about Core files in case of fatal crash: https://rspamd.com/doc/faq.html#how-to-figure-out-why-rspamd-process-crashed
    * Tried ASAN package and obtained the ASAN report (if possible): https://rspamd.com/doc/faq.html#asan-builds 
    * Checked that your issue isn't already filed: https://github.com/issues?utf8=%E2%9C%93&q=is%3Aissue+user%3Arspamd
    * Checked that there is not already an experimental package or master branch

**Describe the bug**
Sorry, I cannot think of a proper concise title for this issue.

I run Rspamd as a milter for Postfix, and today something very peculiar happened. Here are my relevant configuration files:

dmarc.conf:
```
actions = {
  quarantine = "add header";
  reject = "reject";
}
```

actions.conf:
```
add_header = 6;
greylist = 9;
rewrite_subject = null;
```
Now, having been scanned by Rspamd, a spam message triggered the DMARC quarantine policy (according to the logs), yet in the web UI the action "greylist" was shown. However, in reality the message was neither greylisted nor quarantined (no ```X-Spam``` header added). The ```GREYLIST``` symbol was also not present.

![image](https://github.com/rspamd/rspamd/assets/9046118/0e192866-8f8e-4fb7-84bf-11963efeb5ed)

Here's what the log file says:

```
Jul 10 08:28:03 [REDACTED] rspamd[2353]: <1d033f>; proxy; proxy_accept_socket: accepted milter connection from 127.0.0.1 port 33452
Jul 10 08:28:04 [REDACTED] rspamd[2353]: <1d033f>; milter; rspamd_milter_process_command: got connection from 49.51.199.241:37149
Jul 10 08:28:07 [REDACTED] rspamd[2353]: <1d033f>; proxy; rspamd_message_parse: loaded message; id: <D0.67.13791.DE78F546@hs.mta1vrest.cc.prd.sparkpost>; queue-id: <4C4FD2C0063>; size: 49983; checksum: <0418570db6ec092e34ae36fa685ef75f>
Jul 10 08:28:07 [REDACTED] rspamd[2353]: <1d033f>; proxy; rspamd_url_text_extract: got empty text part
Jul 10 08:28:07 [REDACTED] rspamd[2353]: <1d033f>; proxy; rspamd_mime_part_detect_language: detected part language: en
Jul 10 08:28:07 [REDACTED] rspamd[2353]: <1d033f>; proxy; rspamd_spf_maybe_return: not stored SPF record for r239.mail.bose.com (0xa4aa40bbeec59e2b) in LRU cache; flags=4; ttl=0
Jul 10 08:28:07 [REDACTED] rspamd[2353]: <1d033f>; proxy; dkim_module_key_handler: stored DKIM key for app._domainkey.esp1.co in LRU cache for 300 seconds, 66/2000 elements in the cache
Jul 10 08:28:07 [REDACTED] rspamd[2353]: <1d033f>; dkim; rspamd_dkim_check: dkim: bh value mismatch: got 9tvzY3xyHLMRjTTxw64f134OR33N+tGI6KeNZvCx66w=, expected 4yOC7BZM9zb1DFvZUL1+LcX+Tww50WNV7tvJEEEz1U8=; body length 46910->39911; d=esp1.co; s=app
Jul 10 08:28:07 [REDACTED] rspamd[2353]: <1d033f>; proxy; dkim_module_key_handler: cannot get key for domain key1._domainkey.email.myfitnesspal.com: dns request to key1._domainkey.email.myfitnesspal.com failed: no records with this name
Jul 10 08:28:07 [REDACTED] rspamd[2353]: <1d033f>; proxy; rspamd_add_passthrough_result: <D0.67.13791.DE78F546@hs.mta1vrest.cc.prd.sparkpost>: set pre-result to 'add header' *least (no score): 'Action set by DMARC' from dmarc(1)
Jul 10 08:28:07 [REDACTED] rspamd[2353]: <1d033f>; proxy; make_dns_request_task_common: stop resolving on reaching 64 requests
Jul 10 08:28:07 [REDACTED] rspamd[2353]: <1d033f>; proxy; finalize_item: slow rule: SURBL_MULTI(467): 300.00 ms; enable slow timer delay
Jul 10 08:28:07 [REDACTED] rspamd[2353]: <1d033f>; proxy; finalize_item: slow rule: RBL_SEM(436): 387.48 ms
Jul 10 08:28:07 [REDACTED] rspamd[2353]: <1d033f>; proxy; finalize_item: slow rule: SEM_URIBL_UNKNOWN(445): 399.48 ms; enable slow timer delay
Jul 10 08:28:07 [REDACTED] rspamd[2353]: <1d033f>; proxy; finalize_item: slow rule: SEM_URIBL_FRESH15_UNKNOWN(474): 403.48 ms
Jul 10 08:28:08 [REDACTED] rspamd[2353]: <1d033f>; proxy; finalize_item: slow rule: URIBL_MULTI(527): 587.48 ms; enable slow timer delay
Jul 10 08:28:09 [REDACTED] rspamd[2353]: <1d033f>; proxy; finalize_item: slow rule: RSPAMD_URIBL(504): 2360.98 ms; enable slow timer delay
Jul 10 08:28:09 [REDACTED] rspamd[2353]: <1d033f>; proxy; finalize_item: slow rule: PHISHING(424): 2344.98 ms
Jul 10 08:28:09 [REDACTED] rspamd[2353]: <1d033f>; proxy; rspamd_task_write_log: id: <D0.67.13791.DE78F546@hs.mta1vrest.cc.prd.sparkpost>, qid: <4C4FD2C0063>, ip: 49.51.199.241, from: <Rd350qo8RnyNuh1X13H32xTtq@r239.mail.bose.com>, (default: T (greylist): [14.45/15.00] [RBL_NIXSPAM(4.00){49.51.199.241:from;},BAYES_HAM(-3.00){100.00%;},DEPRECATED_RECIPIENT(3.00){[REDACTED];},HFILTER_HOSTNAME_UNKNOWN(2.50){},DMARC_POLICY_QUARANTINE(1.50){myfitnesspal.com : No valid SPF;quarantine;},R_BAD_CTE_7BIT(1.05){7bit;utf8;},DATE_IN_PAST(1.00){41;},EXT_CSS(1.00){},MIME_BASE64_TEXT_BOGUS(1.00){},RDNS_NONE(1.00){},R_DKIM_REJECT(1.00){esp1.co:s=app;},FORGED_SENDER(0.30){hello@email.myfitnesspal.com;Rd350qo8RnyNuh1X13H32xTtq@r239.mail.bose.com;},MIME_BASE64_TEXT(0.10){},MIME_GOOD(-0.10){multipart/alternative;text/plain;},RCVD_NO_TLS_LAST(0.10){},ARC_NA(0.00){},ASN(0.00){asn:132203, ipnet:49.51.196.0/22, country:CN;},DKIM_TRACE(0.00){email.myfitnesspal.com:~;esp1.co:-;},FROM_HAS_DN(0.00){},FROM_NEQ_ENVFROM(0.00){hello@email.myfitnesspal.com;Rd350qo8RnyNuh1X13H32xTtq@r239.mail.bose.com;},MIME_TRACE(0.00){0:+;1:+;2:~;},PREVIOUSLY_DELIVERED(0.00){[REDACTED];},RCPT_COUNT_ONE(0.00){1;},RCVD_COUNT_THREE(0.00){4;},R_DKIM_PERMFAIL(0.00){email.myfitnesspal.com:s=key1;},R_SPF_NA(0.00){no SPF record;},TO_DN_NONE(0.00){},TO_MATCH_ENVRCPT_ALL(0.00){}]), len: 49983, time: 2662.273ms, dns req: 64, digest: <0418570db6ec092e34ae36fa685ef75f>, rcpts: <[REDACTED]>, mime_rcpts: <[REDACTED]>, forced: add header "Action set by DMARC"; score=nan (set by dmarc)
Jul 10 08:28:09 [REDACTED] rspamd[2353]: <1d033f>; proxy; rspamd_protocol_http_reply: regexp statistics: 0 pcre regexps scanned, 2 regexps matched, 175 regexps total, 68 regexps cached, 0B scanned using pcre, 53.53KiB scanned total
```

As a result, the message went to inbox because of no spam header. Rspamd also added the following header (which obviously did not do any good for the IMAP server):

```
X-Rspamd-Pre-Result: action=add header;
    module=dmarc;
    Action set by DMARC
```

### Steps to Reproduce

1. Configure Rspamd as described above.
2. Receive a message to trigger DMARC quarantine policy.

Here is the message in question that triggered the issue (zipped eml, with private information redacted): [wifiverywhere.zip](https://github.com/rspamd/rspamd/files/11998611/wifiverywhere.zip)

**Expected behavior**
The message should either have been marked as spam by means of the add header action, or greylisted. Since the latter has the higher score than the former, and the message spam rating surpassed the greylisting threshold, I think it should've been the latter. But taking no action at all here is by no means appropriate, as it is lets a high-score message directly into the user's inbox.

### Versions

3.5

### Additional Information

None

**Comment by vstakhov** [comment]
The problem is that you have set `greylist` threshold higher than `add header` threshold, so from the perspective of Rspamd `greylist` is a more severe action. Dmarc module sets `least` action, so any more severe action will override it even by score sum. There is no issue from Rspamd side, just a misconfiguration.

**Comment by Player701** [comment]
I've worked around this by disabling the ```add header``` action in ```dmarc.conf``` and instead adding a check for the presence of ```DMARC_POLICY_QUARANTINE``` substring in the ```X-Spamd-Result``` header to the spam filter sieve script.

I still think there is a bug as Rspamd does not currently take _any_ action when configured as originally described (i.e. without the above workaround).

**Comment by vstakhov** [comment]
> So why did greylist not fire then?

Because it is designed not to trigger when any pre-result is set. I'm sorry, but Rspamd is not designed with the assumption that add_header < greylist. Just don't do that, it is quite stupid after all...

**Comment by Player701** [comment]
> Just don't do that, it is quite stupid after all...

I'm sorry, but my experience differs from yours apparently, as I've already said that I find this setup quite effective.

I used to get a lot of spam in the junk folder, but at the same time I didn't want to lower the reject threshold because I'd seen some legitimate senders with badly misconfigured MTAs, so their messages were also sometimes classified as junk by mistake. So I wanted to reduce the amount of spam that got through without having to risk rejecting legitimate messages, and the following options were considered:

- Lowering the reject threshold: bad - bears the risk of rejecting legitimate messages
- Enabling greylisting: bad - default threshold slows down the flow of (mostly) legitimate messages
- Whitelisting misconfigured senders: also bad - everything should be automated

This is not a theory but my own empirical evidence. In the end, I went the current setup as it seems to combine the best of all worlds:

- Reduces the amount of junk mail (spammers _usually_ don't retry)
- Does not reject possibly legitimate mail (legitimate senders _always_ try again)
- Higher greylist threshold does not slow down the flow of primarily legitimate mail

In my opinion, it is much better to filter a potentially legitimate message (i.e. a medium score but still enough to technically consider it spam) into junk than greylist it. The less the spam score, the higher the chance that the sender is legitimate and will thus try again, and in the end greylist would just be an unnecessary delay. Hence, greylist > add header is a valid use case.

I would never have thought I'd been doing it wrong if not for this one special case, and now it turns out it wasn't designed to work this way at all... although, quite surprisingly, it _has_ been working for me for almost 2 years without any trouble.

I suppose I have no option but to use the workaround for now, until something else breaks...
